src/sentry/hybridcloud/rpc/services/caching/service.py

```diff
@@ -0,0 +1,191 @@
+        pass
+
+
+def _set_cache(key: str, value: str, version: int) -> Generator[None, None, bool]:
+    result = cache.add(_versioned_key(key, version), value)
+    yield
+    return result
+
+
+def _versioned_key(key: str, version: int) -> str:
+    return f"{key}.{version}"
+
+
+def _version_model(mode: SiloMode) -> Type[CacheVersionBase]:
+    if mode == SiloMode.REGION:
+        return RegionCacheVersion
+    raise ValueError
+
+
+def _delete_cache(key: str, mode: SiloMode) -> Generator[None, None, int]:
+    version = _version_model(mode).incr_version(key)
+    yield
+    return version
+
+
+def _get_cache(
+    keys: List[str], mode: SiloMode
+) -> Generator[None, None, Mapping[str, Union[str, int]]]:
+    versions = {cv.key: cv.version for cv in _version_model(mode).objects.filter(key__in=keys)}
+    yield
+
+    versioned_keys = [_versioned_key(key, versions.get(key, 0)) for key in keys]
+    existing = cache.get_many(versioned_keys)
+    yield
+    result: Dict[str, Union[str, int]] = {}
+    for k, versioned_key in zip(keys, versioned_keys):
+        if versioned_key in existing:
+            result[k] = existing[versioned_key]
+            continue
+        result[k] = versions.get(k, 0)
+    return result
+
+
+class CacheBackend:
+    """
+    'Exposes' the underlying caching and its versioning system so that tests can fully validate the
+    concurrent consistency of the implementation.  Not intended as a public interface.
+    """
+
+    get_cache = staticmethod(_get_cache)
+    delete_cache = staticmethod(_delete_cache)
+    set_cache = staticmethod(_set_cache)
+
+
+_V = TypeVar("_V")
+
+
+def _consume_generator(g: Generator[None, None, _V]) -> _V:
+    while True:
+        try:
+            g.send(None)
+        except StopIteration as e:
+            return e.value
+
+
+class LocalRegionCachingService(RegionCachingService):
+    def clear_key(self, *, region_name: str, key: str) -> int:
+        return _consume_generator(_delete_cache(key, SiloMode.REGION))
+
+
+class LocalControlCachingService(ControlCachingService):
+    def clear_key(self, *, key: str) -> int:
+        return _consume_generator(_delete_cache(key, SiloMode.CONTROL))
+
+
+_R = TypeVar("_R", bound=pydantic.BaseModel)
+
+
+class SiloCacheBackedCallable(Generic[_R]):
+    silo_mode: SiloMode
+    base_key: str
+    cb: Callable[[int], _R]
+    type: Type[_R]
+
+    def __init__(self, base_key: str, silo_mode: SiloMode, cb: Callable[[int], _R], t: Type[_R]):
+        self.base_key = base_key
+        self.silo_mode = silo_mode
+        self.cb = cb
+        self.type = t
+
+    def __call__(self, args: int) -> _R:
+        if (
+            SiloMode.get_current_mode() != self.silo_mode
+            and SiloMode.get_current_mode() != SiloMode.MONOLITH
+        ):
+            return self.cb(args)
+        return self.get_many([args])[0]
+
+    def key_from(self, args: int) -> str:
+        return f"{self.base_key}:{args}"
```

Comment:
Will other services we need caching on all be able to reduce their `args` to `int`? Are you envisioning this being used only on primary key values?

Reply:
For now, yes.  Expanding beyond int is possible, but it also introduces complexity and performance issues.  This is especially true given that the main cache-clear mechanism is from individual model outboxes.